src/core/reference/include/openvino/reference/fake_convert.hpp

```diff
@@ -0,0 +1,329 @@
+    const unsigned short rne_mask = 1;                                    /* round to nearest even mask */
+    const unsigned short mask_mant = (unsigned short)(0xFFFF << lshift);  // 1111111111111111 -> 1 11111 1111000000
+    constexpr unsigned short grs_bitmask = 0x007F;                        /// 0 00000 0001111111
+    constexpr unsigned short rne_tie = 0x00C0;                            /// 0 00000 0011000000
+
+    __half_t h;
+    for (size_t i = 0; i < count; ++i) {
+        h.f = arg[i];
+        float inval = ngraph::float16(arg[i]);
+        /* flush values below 1-4-3 (offset=4) subnormal range to zero */
+        if (fabs(inval) < 0.001953125)  // 2**-9
+            h.f = 0;
+
+        short exp_h = (short)((h.u & 0x7C00) >> 10) -
+                      15;                   /// 0111110000000000 -> 0000000000011111 - 15 - biased exponent for fp16
+        short sign_h = (h.u & 0x8000);      /// & 1 00000 0000000000
+        short mantissa_h = (h.u & 0x03FF);  /// & 0 00000 1111111111
+        ///(h.u && 0111111111111111) < 0 10010 1110000000 (19326) - ????
+        unsigned short can_round = ((h.u & 0x7FFF) < 0b101111110000000) ? 1 : 0;
+        unsigned short is_naninf = ((h.u & 0x7C00) == 0x7C00) ? 1 : 0;
+
+        int dshift = 0;
+        if (exp_h > 8) {  // too large, set it to NaN or inf
+            is_naninf = 1;
+            if (use_clamp) {
+                exp_h = 8;
+                mantissa_h = 0b0000001100000000;
+            } else {
+                mantissa_h = 0;
+                exp_h = 16;
+            }
+        } else if (exp_h < -9) {  /// -13, -12, -11 for rounding
+            /* flush values below 1-4-3 (offset=4) subnormal range to zero */
+            exp_h = -15;
+            mantissa_h = 0;
+        }
+
+        if (exp_h == 8 && mantissa_h >= 0b0000001100000000) {
+            mantissa_h = 0b0000001100000000;
+        }
+        /* nearest rounding masks, & 0 00000 000 111 1111 - mantissa bits below f8e4m3 (grs) */
+        unsigned short rnmask = (mantissa_h & grs_bitmask);
+        /* & 0 00000 0011000000 - edge between f8e4m3 and fp16 mantissa */
+        unsigned short rnmask_tie = (mantissa_h & rne_tie);
+        if (!is_naninf && can_round && rne_mask) {
+            /* round to nearest even, if rne_mask is enabled */
+            /// rnmask > 0 00000 0001000000(64) or 0 00000 0011000000 - edge bits is 1
+            /// += 0 00000 0010000000
+            mantissa_h += (((rnmask > 0x0040) || (rnmask_tie == rne_tie)) << lshift);
+        }
+        if (exp_h < -6) { /* handle denormals -11, -10, -9, dshift 1, 2, 3 */
+            dshift = (-6 - exp_h);
+            mantissa_h = mantissa_h >> dshift;
+        }
+        mantissa_h &= mask_mant; /* truncation */
+        mantissa_h <<= dshift;
+        mantissa_h += ((exp_h + 15) << 10);
+        h.u = mantissa_h | sign_h;
+        out[i] = h.f;
+    }
+}
+
+template <typename T>
+void apply_scale_shift(T* out,
```

Comment:
If some algorithm applies different function usually better in term of size is do lie

1. Some preparation data
2. select function (function pointer) for algo base on some condition
3. Apply function in algorithm loop.

Instead implements many loops.

Reply:
Added separate `scale_shift` func for those calculations.
As mentioned, changing current approach to a single loop will be considered as a follow up idea to be checked.